pointmatcher/Registrar.h

```diff
@@ -90,80 +90,79 @@ namespace PointMatcherSupport
 		template<typename C>
 		struct GenericClassDescriptor: public ClassDescriptor
 		{
 			virtual std::shared_ptr<Interface> createInstance(const std::string& className, const Parametrizable::Parameters& params) const
 			{
 				std::shared_ptr<C> instance = std::make_shared<C>(params);
-				
+
 				// check that all parameters were set
-				for (BOOST_AUTO(it, params.begin()); it != params.end() ;++it)
+				for (const auto& param : params)
 				{
-					if (instance->parametersUsed.find(it->first) == instance->parametersUsed.end()){
+					if (instance->parametersUsed.find(param.first) == instance->parametersUsed.end()){
 						throw Parametrizable::InvalidParameter(
-							(boost::format("Parameter %1% for module %2% was set but is not used") % it->first % className).str()
+							(boost::format("Parameter %1% for module %2% was set but is not used") % param.first % className).str()
 						);
 					}
 				}
-				
 				return instance;
 			}
+
 			virtual const std::string description() const
 			{
 				return C::description();
 			}
+
 			virtual const Parametrizable::ParametersDoc availableParameters() const
 			{
 				return C::availableParameters();
 			}
 		};
-		
+
 		//! A descriptor for a class C that does not provide any parameter
 		template<typename C>
 		struct GenericClassDescriptorNoParam: public ClassDescriptor
 		{
 			virtual std::shared_ptr<Interface> createInstance(const std::string& className, const Parametrizable::Parameters& params) const
 			{
-				for (BOOST_AUTO(it, params.begin()); it != params.end() ;++it)
+				for (const auto& param : params)
 					throw Parametrizable::InvalidParameter(
-							(boost::format("Parameter %1% was set but module %2% dos not use any parameter") % it->first % className).str()
+							(boost::format("Parameter %1% was set but module %2% dos not use any parameter") % param.first % className).str()
 						);
+
 				return std::make_shared<C>();
 			}
+
 			virtual const std::string description() const
 			{
 				return C::description();
 			}
+
 			virtual const Parametrizable::ParametersDoc availableParameters() const
 			{
 				return Parametrizable::ParametersDoc();
 			}
 		};
-		
+
 	protected:
-		typedef std::map<std::string, ClassDescriptor*> DescriptorMap; //!< descriptors for sub-classes of Interface, indexed by their names
+		typedef std::map<std::string, std::shared_ptr<ClassDescriptor>> DescriptorMap; //!< descriptors for sub-classes of Interface, indexed by their names
 		DescriptorMap classes; //!< known classes that can be constructed
-		
+
 	public:
-		//! Destructor, remove all classes descriptors 
-		~Registrar()
-		{
-			for (BOOST_AUTO(it, classes.begin()); it != classes.end(); ++it)
-				delete it->second;
-		}
 		//! Register a class by storing an instance of a descriptor helper class
-		void reg(const std::string &name, ClassDescriptor* descriptor)
+		void reg(const std::string &name, std::shared_ptr<ClassDescriptor> descriptor)
```

Comment:
Is the descriptor intended to change? Otherwise we could pass a const reference to it.

https://stackoverflow.com/questions/8385457/should-i-pass-a-shared-ptr-by-reference

Reply:
From what I understand, there's no method provided to modify the descriptor, so it seems that it never change, but I can be wrong. I'm not enough familiar with the library to be 100% sure. These changes were made with the help of @simonpierredeschenes.

For the const reference part, I've found [this](https://www.internalpointers.com/post/move-smart-pointers-and-out-functions-modern-c) to have the clearer explanation on how to pass a smart pointer to a function and I would tend to agree with you by passing it by const reference.

Thanks for your suggestion!

@pomerlef and @simonpierredeschenes What do you think?

Reply:
In our usecase, the reg function is only called in the registrar and in that usecase the smart pointer is an rvalue and we are transfering ownership. A const reference can not be used without creating an useless temporary variable. So there are two ways of doing this:
pass by value
```
{
 ...
  reg_by_value("foo", std::make_shared<Foo>()); // Copy constructor
....
}
void reg_by_value(const std::string &name, std::share_ptr<ClassDescriptor> descriptor)
{
  classes.insert(std::make_pair(name, std::move(descriptor))); // move construction
}
```
rvalue reference:
```
{
 ...
  reg_by_rvalue_ref("foo", std::make_shared<Foo>());
....
}
void reg_by_rvalue_ref(const std::string& name, std::share_ptr<ClassDescriptor>&& descriptor)
{
  classes.insert(std::make_pair(name, descriptor)); // move construction
}
```
The second solution prevent a copy. However, the interface now required a rvalue reference, which would be a problem if this function is used outside of the library.

Reply:
Makes sense. Thank you :+1:

Reply:
Thank you @PhiBabin, these are really interesting solutions. I've found in [this article](https://herbsutter.com/2013/06/05/gotw-91-solution-smart-pointer-parameters/) that passing a temporary shared pointer as parameter (which is the case here), it will be move construct. So, I think that it would make your first solution look like this?

```cpp
{
 ...
  reg_by_value("foo", std::make_shared<Foo>()); // move construction
 ...
}
void reg_by_value(const std::string &name, std::share_ptr<ClassDescriptor> descriptor)
{
  classes.insert(std::make_pair(name, descriptor));
}
```
Tell me what you think and feel free to correct me... I'm still learning the C++ subtleties! :)

Reply:
You are right, in C++11, in this example when we use `reg_by_value` we will have two move constructions (not 1 copy + 1 move construction), where as `reg_by_rvalue_ref` will have a single move construction.

Reply:
Thanks @PhiBabin, let's go for this then! 👍🏻